File: tagbag/tag_orderer.py
from .tag_base import TagBase
from .tag_helper import TagHelper


from pathlib import Path
from typing import Dict, Iterable, List, Sequence, Tuple


class TagOrderer(TagBase):
    """
    Order tags per line using a global position score learned from all files.

    For each item at index j among n items on a line, the contribution is j * (j / n).
    The running average is updated across all lines.
    """

    @staticmethod
    def execute(args: Sequence[str]) -> None:
        """
        args[0]: directory path
        """
        file_names = TagBase.get_tag_files(args[0])
        tokens: Dict[str, Tuple[int, float]] = {}
        TagOrderer._count_tags_in_files(file_names, tokens)
        TagOrderer._order(file_names, tokens)

    @staticmethod
    def _count_tags_in_files(file_names: Iterable[str], tokens: Dict[str, Tuple[int, float]]) -> None:
        for name in file_names:
            TagOrderer._count_tags_in_file(Path(name), tokens)

    @staticmethod
    def _count_tags_in_file(file_path: Path, tokens: Dict[str, Tuple[int, float]]) -> None:
        for line in file_path.read_text(encoding="utf-8").splitlines():
            items = [t.strip() for t in TagHelper.split_into_tags(line) if t.strip()]
            n = len(items)
            if n == 0:
                continue
            for j, item in enumerate(items):
                current_position = j * (j / float(n))
                if item in tokens:
                    count, avg_pos = tokens[item]
                    new_count = count + 1
                    new_pos = (current_position / new_count) + (avg_pos * ((new_count - 1) / float(new_count)))
                    tokens[item] = (new_count, new_pos)
                else:
                    tokens[item] = (1, current_position)

    @staticmethod
    def _order(file_names: Iterable[str], tokens: Dict[str, Tuple[int, float]]) -> None:
        for name in file_names:
            p = Path(name)
            lines = p.read_text(encoding="utf-8").splitlines()
            out_lines: List[str] = []

            for line in lines:
                items = [t.strip() for t in TagHelper.split_into_tags(line) if t.strip()]
                scored: List[Tuple[str, float]] = []
                for item in items:
                    if item in tokens:
                        scored.append((item, tokens[item][1]))
                scored.sort(key=lambda x: x[1])
                out_lines.append(", ".join(token for token, _ in scored))

            p.write_text("\n".join(out_lines), encoding="utf-8")

File: tagbag/TagOrderer.cs
namespace NhlStenden.TagBag;

/// <summary>
/// Represents a class that orders tags in files based on their frequency and position.
/// </summary>
public class TagOrderer : TagBase
{
    /// <summary>
    /// Executes the tag ordering process on the specified files.
    /// </summary>
    /// <param name="args">A list of arguments where the first argument is the directory path.</param>
    public static void Execute(IList<string> args)
    {
        var fileNames = GetTagFiles(args[0]);
        var tokens = new Dictionary<string, (int Count, float Position)>();
        CountTagsInFiles(fileNames, tokens);

        Order(fileNames, tokens);
    }

    /// <summary>
    /// Counts the tags in the specified files and updates the token dictionary.
    /// </summary>
    /// <param name="fileNames">An array of file names to process.</param>
    /// <param name="tokens">A dictionary to store the count and position of each tag.</param>
    static void CountTagsInFiles(string[] fileNames, Dictionary<string, (int Count, float Position)> tokens)
    {
        foreach (var fileName in fileNames)
            CountTagsInFile(fileName, tokens);
    }

    /// <summary>
    /// Counts the tags in the specified file and updates the token dictionary.
    /// </summary>
    /// <param name="fileName">The name of the file to process.</param>
    /// <param name="tokens">A dictionary to store the count and position of each tag.</param>
    static void CountTagsInFile(string fileName, IDictionary<string, (int Count, float Position)> tokens)
    {
        var lines = File.ReadAllLines(fileName).ToList();

        for (var i = 0; i < lines.Count; i++)
        {
            var line = lines[i];

            var items = TagHelper.SplitIntoTags(line);
            for (var j = 0; j < items.Count; j++)
            {
                var item = items[j];
                var currentPosition = j * (j / (float)items.Count);
                if (tokens.TryGetValue(item, out var counter))
                {
                    var count = counter.Count + 1;
                    var newPosition = (currentPosition / count) + (counter.Position * ((count - 1) / ((float)count)));
                    tokens[item] = (count, newPosition);
                }
                else
                {
                    tokens[item] = (1, currentPosition);
                }
            }
        }
    }

    /// <summary>
    /// Orders the tags in the specified files based on their frequency and position.
    /// </summary>
    /// <param name="fileNames">An array of file names to process.</param>
    /// <param name="tokens">A dictionary containing the count and position of each tag.</param>
    static void Order(string[] fileNames, Dictionary<string, (int Count, float Position)> tokens)
    {
        foreach (var fileName in fileNames)
        {
            var lines = File.ReadAllLines(fileName).ToList();

            for (var i = 0; i < lines.Count; i++)
            {
                var line = lines[i];

                var items = TagHelper.SplitIntoTags(line);
                var newTokes = new List<(string Token, float Position)>();
                for (var j = 0; j < items.Count; j++)
                {
                    var item = items[j];
                    if (!tokens.TryGetValue(item, out var counter))
                        continue;
                    newTokes.Add((item, counter.Position));
                }
                newTokes = newTokes.OrderBy(x => x.Position).ToList();
                line = string.Join(", ", newTokes.Select(x => x.Token));
                lines[i] = line;
            }
            File.WriteAllLines(fileName, lines);
        }
    }
}